File: discord-bot-fleet/src/tools/memory.ts
// memory — persistent per-bot memory store. Each bot has its own JSON file.
// Safe: path is scoped, no traversal possible. Size-capped to prevent disk abuse.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import { join } from 'path';
import type { LLMTool } from '../llm.js';
import type { ToolContext } from './index.js';

const DATA_DIR = process.env.DATA_DIR || join(process.cwd(), 'data');
const MEMORY_DIR = join(DATA_DIR, 'memory');
if (!existsSync(MEMORY_DIR)) mkdirSync(MEMORY_DIR, { recursive: true });

// Caps: prevent disk abuse from runaway bots
const MAX_KEYS = 500;
const MAX_VALUE_BYTES = 16 * 1024; // 16KB per value
const MAX_FILE_BYTES = 5 * 1024 * 1024; // 5MB total per bot

function memoryFile(botId: string): string {
  // Sanitize: only alphanumerics allowed (bot IDs are nanoid)
  const safe = botId.replace(/[^a-zA-Z0-9_-]/g, '');
  return join(MEMORY_DIR, `${safe}.json`);
}

function loadMemory(botId: string): Record<string, string> {
  const f = memoryFile(botId);
  if (!existsSync(f)) return {};
  try {
    return JSON.parse(readFileSync(f, 'utf8'));
  } catch {
    return {};
  }
}

function saveMemory(botId: string, mem: Record<string, string>) {
  const f = memoryFile(botId);
  // Hard cap: don't let a single bot's memory file exceed 5MB
  const serialized = JSON.stringify(mem);
  if (serialized.length > MAX_FILE_BYTES) {
    throw new Error(`memory file would exceed ${MAX_FILE_BYTES} bytes — too many keys/values`);
  }
  writeFileSync(f, serialized);
}

export const memoryReadTool: LLMTool = {
  type: 'function',
  function: {
    name: 'memory_read',
    description: 'Read a value from your persistent memory by key, or list all keys if no key is given. Use this to recall things about users, ongoing topics, or anything you decided to remember.',
    parameters: {
      type: 'object',
      properties: {
        key: { type: 'string', description: 'Memory key to read. If omitted, returns a list of all keys.' },
      },
    },
  },
};

export async function memoryReadHandler(
  args: { key?: string },
  ctx: ToolContext
): Promise<string> {
  const mem = loadMemory(ctx.botConfig.id);
  if (!args.key) {
    return JSON.stringify({ keys: Object.keys(mem), count: Object.keys(mem).length });
  }
  if (args.key in mem) {
    return JSON.stringify({ key: args.key, value: mem[args.key] });
  }
  return JSON.stringify({ key: args.key, value: null, note: 'key not found' });
}

export const memoryWriteTool: LLMTool = {
  type: 'function',
  function: {
    name: 'memory_write',
    description: 'Write a key/value pair to your persistent memory. Use this to remember facts about users, ongoing conversations, decisions, or anything you want to recall later. Values are limited to 16KB; max 500 keys per bot.',
    parameters: {
      type: 'object',
      properties: {
        key: { type: 'string', description: 'Memory key (use snake_case, e.g. "user_bob_favorite_lang"). Max 100 chars.' },
        value: { type: 'string', description: 'Value to store. Max 16KB.' },
      },
      required: ['key', 'value'],
    },
  },
};

export async function memoryWriteHandler(
  args: { key: string; value: string },
  ctx: ToolContext
): Promise<string> {
  // Validate key
  if (typeof args.key !== 'string' || args.key.length === 0) {
    return JSON.stringify({ error: 'key must be a non-empty string' });
  }
  if (args.key.length > 100) {
    return JSON.stringify({ error: `key too long (${args.key.length} > 100 chars)` });
  }
  // Validate value size
  if (typeof args.value !== 'string') {
    return JSON.stringify({ error: 'value must be a string' });
  }
  const valueBytes = Buffer.byteLength(args.value, 'utf8');
  if (valueBytes > MAX_VALUE_BYTES) {
    return JSON.stringify({ error: `value too large (${valueBytes} bytes > ${MAX_VALUE_BYTES}). Truncate or split into multiple keys.` });
  }

  const mem = loadMemory(ctx.botConfig.id);

  // Enforce max keys (only when adding a new key)
  if (!(args.key in mem) && Object.keys(mem).length >= MAX_KEYS) {
    return JSON.stringify({ error: `memory full: ${MAX_KEYS} keys max. Delete some keys first.` });
  }

  mem[args.key] = args.value;
  try {
    saveMemory(ctx.botConfig.id, mem);
  } catch (e: unknown) {
    return JSON.stringify({ error: (e as Error).message });
  }
  return JSON.stringify({ ok: true, key: args.key, size_bytes: valueBytes });
}

File: discord-bot-fleet/bot-python/tools/memory.py
"""memory — persistent per-bot memory. Same JSON file as Node, size-capped."""

import json
import os
import re
from typing import Dict

DATA_DIR = os.environ.get('DATA_DIR', os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', 'data'))
MEMORY_DIR = os.path.join(DATA_DIR, 'memory')
os.makedirs(MEMORY_DIR, exist_ok=True)

MAX_KEYS = 500
MAX_VALUE_BYTES = 16 * 1024
MAX_FILE_BYTES = 5 * 1024 * 1024

MEM_PERSONA_READ = {
    'type': 'function',
    'function': {
        'name': 'memory_read',
        'description': 'Read a value from your persistent memory by key, or list all keys if no key is given. Use this to recall things about users, ongoing topics, or anything you decided to remember.',
        'parameters': {
            'type': 'object',
            'properties': {
                'key': {'type': 'string', 'description': 'Memory key to read. If omitted, returns a list of all keys.'},
            },
        },
    },
}

MEM_PERSONA_WRITE = {
    'type': 'function',
    'function': {
        'name': 'memory_write',
        'description': 'Write a key/value pair to your persistent memory. Use this to remember facts about users, ongoing conversations, decisions, or anything you want to recall later. Values are limited to 16KB; max 500 keys per bot.',
        'parameters': {
            'type': 'object',
            'properties': {
                'key': {'type': 'string', 'description': 'Memory key (use snake_case, e.g. "user_bob_favorite_lang"). Max 100 chars.'},
                'value': {'type': 'string', 'description': 'Value to store. Max 16KB.'},
            },
            'required': ['key', 'value'],
        },
    },
}

# Export under the names the registry expects
MEMORY_READ_TOOL = MEM_PERSONA_READ
MEMORY_WRITE_TOOL = MEM_PERSONA_WRITE


def _memory_file(bot_id: str) -> str:
    safe = re.sub(r'[^a-zA-Z0-9_-]', '', bot_id)
    return os.path.join(MEMORY_DIR, f'{safe}.json')


def _load_memory(bot_id: str) -> Dict[str, str]:
    f = _memory_file(bot_id)
    if not os.path.exists(f):
        return {}
    try:
        with open(f, 'r') as fh:
            return json.load(fh)
    except Exception:
        return {}


def _save_memory(bot_id: str, mem: Dict[str, str]) -> None:
    f = _memory_file(bot_id)
    serialized = json.dumps(mem)
    if len(serialized) > MAX_FILE_BYTES:
        raise ValueError(f'memory file would exceed {MAX_FILE_BYTES} bytes — too many keys/values')
    with open(f, 'w') as fh:
        fh.write(serialized)


async def memory_read_handler(args, ctx):
    key = args.get('key')
    mem = _load_memory(ctx.bot_config['id'])
    if not key:
        return json.dumps({'keys': list(mem.keys()), 'count': len(mem)})
    if key in mem:
        return json.dumps({'key': key, 'value': mem[key]})
    return json.dumps({'key': key, 'value': None, 'note': 'key not found'})


async def memory_write_handler(args, ctx):
    key = args.get('key', '')
    value = args.get('value', '')
    if not isinstance(key, str) or not key:
        return json.dumps({'error': 'key must be a non-empty string'})
    if len(key) > 100:
        return json.dumps({'error': f'key too long ({len(key)} > 100 chars)'})
    if not isinstance(value, str):
        return json.dumps({'error': 'value must be a string'})
    value_bytes = len(value.encode('utf-8'))
    if value_bytes > MAX_VALUE_BYTES:
        return json.dumps({'error': f'value too large ({value_bytes} bytes > {MAX_VALUE_BYTES}). Truncate or split into multiple keys.'})
    mem = _load_memory(ctx.bot_config['id'])
    if key not in mem and len(mem) >= MAX_KEYS:
        return json.dumps({'error': f'memory full: {MAX_KEYS} keys max. Delete some keys first.'})
    mem[key] = value
    try:
        _save_memory(ctx.bot_config['id'], mem)
    except Exception as e:
        return json.dumps({'error': str(e)})
    return json.dumps({'ok': True, 'key': key, 'size_bytes': value_bytes})
